### ISSUE TITLE
Fix issues with tracing

### ISSUE THREAD

**Issue by sylwiaszunejko** [scylladb/gocql#227]
In gocql there is no mechanism to retry the session detail fetch, and there is a need for it because tracing details are stored asynchronously. Other drivers have it:
- in python-driver there is a mechanism to retry the session detail fetch, because tracing details are stored asynchronously, if the trace is still not available after max_wait seconds, TraceUnavailable will be raised - https://github.com/scylladb/python-driver/blob/8a4387ae3f36522cc10841c607abb8182c6f8286/cassandra/query.py#L981-L994
- in rust-driver there is a tracing_info_fetch_attempts field that defines how many times the driver will try to fetch tracing information with sleeps in between https://github.com/scylladb/scylla-rust-driver/blob/36d5edd3a06a6808ce19dd594ea4904c8b48c0c2/scylla/src/transport/session.rs#L1411-L1422
Another problem that needs to be fixed is reading of tracing info should be done asynchronously, otherwise enabling tracing will lead to query processing slow down.
Also `traceWriter` is actually not exactly useful, it just write custom formatted messages into a buffer.
So, probably we should abandon `traceWriter` and stop using it even in tests. Instead come up with an API that will solve things easily and without running tracing subroutines.

**Comment by sylwiaszunejko** [comment]
I would like to start a discussion about this. Refs: https://github.com/scylladb/gocql/issues/215#issuecomment-2268901657

**Comment by sylwiaszunejko** [comment]
For the first problem I was thinking about adding similar logic as in rust-driver: https://github.com/scylladb/gocql/compare/master...sylwiaszunejko:gocql:flaky_tablet_test

**Comment by dkropachev** [comment]
> For the first problem I was thinking about adding similar logic as in rust-driver: [master...sylwiaszunejko:gocql:flaky_tablet_test](https://github.com/scylladb/gocql/compare/master...sylwiaszunejko:gocql:flaky_tablet_test)

It does not consider that data could be partially written.
There is distinctive message in `system_traces.events` activity=` Request complete` that signals that all tracing data is already written.
We should wait for it to apear otherwise continue reading, including coordinator durations.

**Comment by dkropachev** [comment]
`traceWriter` is not exactly usefull, it just write custom formatted messages into a buffer.

Having that said my proposal is following:
1. Mark current `traceWriter` as depricated and stop using it
2. Create a new tracer, with following interface
```
func (t *tracer) Trace(traceId []byte)
func (t *tracer) AllTraceIDs() [][]byte
func (t *tracer) GetCoordinatorTime(traceId []byte) (coordinator, duration time.Duration, err error)
func (t *tracer) GetActivity(traceId []byte) (timestamp time.Time, activity string, source string, elapsed int, thread string, err error)
```

The idea of how to use it is following:
```
		trace := NewTraceWriter(session, buf)
		err = session.Query(`SELECT pk, ck, v FROM test1.table1 WHERE pk = ?;`, i).WithContext(ctx).Consistency(One).Trace(trace).Scan(&pk, &ck, &v)
		if err != nil {
			t.Fatal(err)
		}
		for _, traceID := range trace.AllTraceIDs() {
				coordinator, d, err := trace.GetCoordinatorTime(traceID)
				.....
				timestamp, activity, source, elapsed, thread, err := trace.GetActivity(traceID)
				.....
		}
```

As bonus we could split trace id by types: prepared statement, query, batch_query.

**Comment by sylwiaszunejko** [comment]
@wprzytula @Lorak-mmk maybe you know more details on how these problems are resolved in e.g. scylla-rust-driver and have any suggestion which approach should we choose in gocql?

**Comment by sylwiaszunejko** [comment]
I am working on a new API with the interface @dkropachev  proposed, but I think it would be still good to have the old API actually working properly with retries, so I created this PR: #233

**Comment by sylwiaszunejko** [comment]
> `traceWriter` is not exactly usefull, it just write custom formatted messages into a buffer.
> 
> Having that said my proposal is following:
> 
>     1. Mark current `traceWriter` as depricated and stop using it
> 
>     2. Create a new tracer, with following interface
> 
> 
> ```
> func (t *tracer) Trace(traceId []byte)
> func (t *tracer) AllTraceIDs() [][]byte
> func (t *tracer) IsReady(traceId []byte) (bool, error)
> func (t *tracer) GetCoordinatorTime(traceId []byte) (coordinator, duration time.Duration)
> func (t *tracer) GetActivity(traceId []byte) (timestamp time.Time, activity string, source string, elapsed int, thread string)
> ```
> 
> The idea of how to use it is following:
> 
> ```
> 		trace := NewTracer(session)
> 		err = session.Query(`SELECT pk, ck, v FROM test1.table1 WHERE pk = ?;`, i).WithContext(ctx).Consistency(One).Trace(trace).Scan(&pk, &ck, &v)
> 		if err != nil {
> 			t.Fatal(err)
> 		}
> 		for _, traceID := range trace.AllTraceIDs() {
> 				var (
> 					isReady bool
> 					err error
> 				)
> 				for !isReady {
> 					isReady, err = trace.IsReady(traceID)
> 					if err != nil {
> 						return err
> 					}
> 				}
> 				timestamp, activity, source, elapsed, thread := trace.GetActivity(traceID)
> 				coordinator, d, err := trace.GetCoordinatorTime(traceID)
> 				.....
> 		}
> ```

I have a question about this design. Shouldn't we have `GetActivities` returning list of all activities for given `traceId`? Something like that:
```
type TraceEntry struct {
	Timestamp time.Time
	Activity  string
	Source    string
	Elapsed   int
	Thread    string
}

func (t *tracer) GetActivities(traceId []byte) ([]TraceEntry, error)
```

**Comment by dkropachev** [comment]
> ```
> type TraceEntry struct {
> 	Timestamp time.Time
> 	Activity  string
> 	Source    string
> 	Elapsed   int
> 	Thread    string
> }
> 
> func (t *tracer) GetActivities(traceId []byte) ([]TraceEntry, error)
> ```

You are correct, same goes for `GetCoordinatorTime`, it should return a list.
